docs/dificil/rmq.md

```diff
@@ -1,18 +1,827 @@
+
+```cpp
+#include <iostream>
+
+const int MAXN = 100'000;
+const int MAXVAL = 1 << 30;
+const int MAXLOG = 17;
+
+int n;
+unsigned int x, y, z;
+
+int max(int a, int b) {
+    return a > b ? a : b;
+}
+
+struct SparseTable {
+    int spt[MAXLOG][MAXN], lg2[MAXN + 1];
+
+    void init(int n) {
+        int i, j;
+        for (i = 2; i <= n; i++) {
+            lg2[i] = 1 + lg2[i >> 1];
+        }
+        for (i = 0; i <= lg2[n]; i++) {
+            for (j = 0; j < n; j++) {
+                spt[i][j] = 0;
+            }
+        }
+    }
+
+    void update(int st, int dr, int val) {
+        int lg = lg2[dr - st + 1], dr_idx;
+        spt[lg][st] = max(spt[lg][st], val);
+        dr_idx = dr - (1 << lg) + 1;
+        spt[lg][dr_idx] = max(spt[lg][dr_idx], val);
+    }
+
+    void pullResults() {
+        int i, j, new_j;
+        for (i = lg2[n]; i > 0; i--) {
+            for (j = 0; j + (1 << i) <= n; j++) {
+                spt[i - 1][j] = max(spt[i - 1][j], spt[i][j]);
+                new_j = j + (1 << (i - 1));
+                spt[i - 1][new_j] = max(spt[i - 1][new_j], spt[i][j]);
+            }
+        }
+    }
+
+    void writeAnswer() {
+        int i;
+        unsigned long long answer;
+        pullResults();
+        answer = 0;
+        for (i = 0; i < n; i++) {
+            answer ^= 1LL * (i + 1) * spt[0][i];
+        }
+        std::cout << answer << "\n";
+    }
+} rmq;
+
+unsigned int nextValue() {
+    unsigned int w;
+    x ^= x << 11;
+    x ^= x >> 4;
+    x ^= x << 5;
+    x ^= x >> 14;
+    w = x ^ y ^ z;
+    x = y;
+    y = z;
+    z = w;
+    return z;
+}
+
+void processUpdates() {
+    int q, i, st, dr, val, aux;
+    std::cin >> n >> q >> x >> y >> z;
+    rmq.init(n);
+    for (i = 0; i < q; i++) {
+        st = nextValue() % n;
+        dr = nextValue() % n;
+        if (st > dr) {
+            aux = st;
+            st = dr;
+            dr = aux;
+        }
+        val = nextValue() % MAXVAL;
+        rmq.update(st, dr, val);
+    }
+}
+
+int main() {
+    int t;
+    std::cin >> t;
+    while (t--) {
+        processUpdates();
+        rmq.writeAnswer();
+    }
+    return 0;
+}
+```
+
+## RMQ sau AINT?
+
+Arborii de intervale (AINT) pot face tot ce poate face RMQ, dar haideți să comparăm aceste două structuri de date.
+
+### Comparație în funcție de timp
+
+RMQ este precalculat, la început, în $O(n \log n)$, iar AINT în $O(n)$. Însă, RMQ are complexitate $O(1)$ per query, În timp ce AINT are $O(\log n)$ per query.
+
+### Comparație în funcție de memorie
+
+RMQ folosește $O(n \log n)$ memorie, iar AINT folosește $O(n)$ memorie.
+
+### Alte precizări
+
+AINT poate rezolva și alte probleme pe care RMQ nu le poate rezolva. La astfel de probleme, putem folosi un Sparse Table normal, rezultând aceeași complexitate ca la AINT la query-uri. Acest lucru înseamnă că AINT este mai folositor în acest caz, chiar dacă este mai greu de implementat.
```

Comment:
poti detalia eventual, pentru ca la aint poti trata si update-urile in timp ce rmq nu poate